modules/virtual_network_gateway/main.tf

```diff
@@ -1,86 +1,119 @@
+# https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/public_ip
+resource "azurerm_public_ip" "this" {
+  for_each = { for ip_configuration in var.ip_configuration : ip_configuration.name => ip_configuration.name if ip_configuration.create_public_ip }
```

Comment:
I'm not saying you code is wrong :) just an idea I'd like to share:
```hcl
toset([ for ipc in var.ip_configuration : ipc.name if ipc.create_public_ip ])
```

Reply:
done